torchmultimodal/models/gpt.py

```diff
@@ -28,12 +28,221 @@ class TransformerLayerOutput(NamedTuple):
     past_key_values: Optional[Dict[str, Tensor]] = None
 
 
-class MultimodalTransformerDecoder(nn.Module):
-    """Extends the transformer decoder of GPT (Generative Pre-Training) model for cross-modality generation.
+class MultimodalGPTOutput(NamedTuple):
+    decoder_output: TransformerDecoderOutput
+    logits: Tensor
+
+
+class MultimodalGPT(nn.Module):
+    """Extends the GPT (Generative Pre-Training) model for cross-modality generation.
 
-    This module implements the transformer decoder of GPT model for generation of one modality given another
+    This module implements the GPT model for generation of one modality given another
     following the paper `"Improving Language Understanding by Generative Pre-Training
     "<https://cdn.openai.com/research-covers/language-unsupervised/language_understanding_paper.pdf>`_.
+
+    Attributes:
+        d_model (int): Embedding dimension of the transformer decoder.
+        num_tokens (int): Total number of tokens for the input and output modalities.
+        in_tokenizer (nn.Module): Tokenizer for the input modality.
+        out_tokenizer (nn.Module): Tokenizer for the output modality.
+        mm_decoder (nn.Module): Multimodal transformer decoder.
+
+    Args:
+        in_modality (Tensor, optional): Tensor of dimension ``(b, in_seq_len, c)`` containing tokenized
+            embeddings for the input modality. Defaults to ``None``.
+        out_modality (Tensor, optional): Tensor of dimension ``(b, out_seq_len, c')`` containing tokenized
+            embeddings for the output modality. Defaults to ``None``.
+        in_pos_ids (Tensor, optional): Tensor of dimension ``(b, in_seq_len)`` containing indices for the
+            input modality position embeddings. Defaults to ``None``.
+        out_pos_ids (Tensor, optional): Tensor of dimension ``(b, out_seq_len)`` containing indices for the
+            output modality position embeddings. Defaults to ``None``.
+        attn_mask (Tensor, optional): Tensor of dimension ``(b, total_seq_len, total_seq_len)``. Contains 1s for
```

Comment:
that's right. We are doing self attention here so query and key sequences are the same. Will edit.

Reply:
Ended up not doing this. 

Because self attention mask requires `1`s only in the lower triangle of the matrix ``(query_length, key_length)`` (due to causal) so it cannot be broadcasted from either query or key dim (if only one is given).

- The two dimensional case is only for padding masks
- The three dimensional case is of self attention masks

https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/utils/attention.py#L30-L38